### ISSUE TITLE
Help needed to integrate Vorple save/restore in Electron

### ISSUE THREAD

**Issue by Natrium729** [vorple/vorple#35]
I'm writing a boilerplate to easily release a Vorple project with Electron:

https://bitbucket.org/Natrium729/vorplectron

I could manage most of its features, but I can't manage to use [Electron native dialogs](https://electronjs.org/docs/api/dialog#dialogshowopendialogbrowserwindow-options-callback) when saving and restoring the game.

Following is an explanation of what I tried.

Since I didn't want to delve into Vorple's low-level working, I thought I would use [an extension I wrote](https://bitbucket.org/Natrium729/extensions-inform-7/src/default/Save%20File%20Control.i7x), "Save File Control", that make it possible to save the game's state into an Inform external file. ([It is in fact code that comes from *Kerkerkruip*.](https://github.com/i7/kerkerkruip/blob/master/Kerkerkruip.materials/Extensions/Victor%20Gijsbers/Kerkerkruip%20Permadeath.i7x))

The plan was this. Within the game, I save the game state into an Inform external file and I use Emscripten file system to read its data, that I send to Electron; Electron then opens the dialog and write the data to the selected location using Node.js file system. I do that the other way around when restoring the game.

This work fine when using normal external files (with text inside), but not with an external holding a save: restoring a file that was saved on the hard disk fails. Using the Save File Control extension directly within the story (such as in its "Permadeath" exemple) without messing with Electron works fine.

If you want to see the code that I wrote:

* [The Inform side is in this extension.](https://bitbucket.org/Natrium729/extensions-inform-7/src/26989f3b7de70f77439bd1f90fc610df7962ef10/Vorplectron.i7x#lines-86)
* [The Electron side is here.](https://bitbucket.org/Natrium729/vorplectron/src/4a217a873169aa58d15e97f397cdb661fb6f9627/listeners.js#lines-47)

So I'm asking what my other options to use Electron native dialogs to save and restore the game are. It would be great if there were a way to ask Vorple to return an object (or a string or whatever) containing the encoded game state and a way to send back an encoded state to Vorple.

It's not that urgent, but a friend of mine will soon release a preview of his Vorple game. While not having the possibility to save and restore in the demo is not critical, this is the only blocking issue remaining. So even a temporary workaround would be appreciated.

Thanks!

**Comment by Natrium729** [comment]
Well, I read and open the file in the "binary" encoding, both in Node.js and Emscripten, so I don't really know. It may as well just be a stupid bug from my part.

I have seen that the bug for saving is solved in Vorple's next version, but will it be easy to hook custom dialogs into Vorple "native" saving/restoring/transcripting? (And not only Electron's ones: it may be useful to be able to hook any system, even in pure JS/HTML. It would be useful when I'll try Vorple on mobile with Capacitor.)

Anyway, it's not a big problem for the moment. I'd be thankful if you take a look, but I think it's better for the moment that you concentrate on the next release of Vorple.

**Comment by curiousdannii** [comment]
Is it using NodeFS? If it isn't I suspect that's because it tests for browsers first and node second, so it thinks electron is a browser environment.

**Comment by Natrium729** [comment]
Node.js integration is disabled in the web page, so it can't use Node.js `fs`. Instead, the web page sends a message containing the file data (read with Emscripten) to Electron's main process (the "server side", which has no access to Emscripten) and Electron uses this data with Node.js file system.

As a matter of fact, everything works with plain default Inform external files. The game can write data to a file, the user downloads it, then the game writes something else and the users uploads the file he downloaded earlier. At this point, if the game reads the file, the uploaded content is shown instead of the most recent write.

It's only with the binary files created by the Save File Control Extension that it does not work.

However, If I open the story in a regular interpreter (Lectrote) and load the downloaded file created by Save File Controle the regular way (by typing RESTORE), the game state is read correctly, but with a bug that causes the prompt to disappear. (We can still type commands, but no prompt is shown.)

So for the moment, I just haven't got any idea on what is happening. 🙁

If someone has, that would be nice, but as I said, it may be better to wait for the next release of Vorple.

**Comment by Natrium729** [comment]
So the latest release fixed the saving/restoring bug. Does it add an easy way to hook a custom dialog?

I dug a bit in haven's source code, and I think I could hook a restore dialog in haven/file.js, in the function `prompt`. (I would replace `window.prompt` with Electron's `dialog.showOpenDialog`, read the data at the opened path, write it with Emscripten FS to `/gamedata_${checksum}/savename.glksave`.)

However, it wouldn't work when saving, because then the dialog has to be opened *after* the save has occured (so that we can read the data and write it to the user's computer.) Another problem is that it is not possible to detect what kind of file is being processed (a save, a transcript, etc.).

We would need another hook at the end of `glk_fileref_create_by_prompt` in cheapglk/cgfref.c, but then there would be a lot to refactor, I think.

Or maybe I'm just looking too far and an easier way exists?

(By the way, the documentation for the function `prompt` in haven/file.js is not correct. It states that the argument `why` has to be one of "for command recording", "for command playback", "to restore", "to save" or "to begin transcription (or printer name)", which is not the case since in cheapglk/cgfref.c, the argument used in `EM_ASM` is `prompt2`, which can only be "to store" or "to load".)

**Comment by juhana** [comment]
This is a bit hacky, but you'd need to delay the save action from Inform so that you'll first trigger a "fake" save that gets the filename from the user, and then trigger the "real" save that creates the file. Once that's done you can pull the save file from the virtual filesystem.

You'll first need to replace the `haven.file.prompt` method so that it doesn't call `window.prompt`:

```js
let filename = "";

haven.file = {
    ...haven.file,
    prompt: function( why ) {
        // pass the filename to the engine
        Module.ccall(
            'haven_getfilename',
            'null',
            [ 'string' ],
            [ filename + '\n' ]
        );
    }
};
```

Then in Inform you'll need to redirect the save action triggered by the player to run the custom save method instead. That should set the `filename` variable's value.

```inform7
Understand the command "__save" as "save".

Check saving the game:
	if the player's command matches "__save":
		continue the action;
	otherwise:
		execute JavaScript command "// whatever triggers the custom save action";
		stop the action.

Report saving the game:
	execute JavaScript command "// the file has now been saved to the virtual filesystem as filename+'.glksave'".
```

Once you have the real file's information, trigger the actual save mechanism from JS with `vorple.prompt.queueCommand("__save", true);` and finish by calling a JS function in a "report saving the game" rule when the virtual save file exists.

The `why` argument variations come from the Hugo engine. You can use it to check if it's a save or a restore action.

**Comment by Natrium729** [comment]
Yeah, that's clever ! I'll report when I'll have the time to try it.

I have a few questions.

* Is this the "official" way of doing it, or Vorple will introduce a less "hacky" way some day?
* Where should I replace `haven.file`? In a Vorple Interface Setup rule?
* If the process is synchronous, I suppose I don't need the 2-stage saving, with `save` and `__save`?

Thanks a lot!

**Comment by sanyayadav94** [comment]
Is save/restore supposed to be working on the new version of vorple.  I am using the latest version, but still always get the error 'Restore failed' when I try to implement it in my game. Any ideas to troubleshoot?

**Comment by sanyayadav94** [comment]
Additionally, I am really having trouble accessing transcripts. I am trying to find the API as referenced here https://vorple-if.com/docs/filesystem.html, for more details, but the link seems to be broken.

**Comment by juhana** [comment]
There was a bug in version 3.2.6 that prevented save files from working. It's now fixed in 3.2.7, hopefully that fixes your problem as well!

API links have also been fixed in the documentation (should have been https://vorple-if.com/api/module-file.html.) You can also get there with the link in the main menu.
